### PR TITLE
Add code guidelines

### PR DESCRIPTION
**Pull request by mks-m**
will probably ignite some heated debates but whatever. guidelines are needed and whatever we agree on should live in this doc/CODING_GUIDELINES.md file.

### CODE REVIEW

File: bin/api [hunk 1]
```diff
@@ -6,4 +6,4 @@
 Buttercoin = require('../lib/buttercoin')
 butter = new Buttercoin();
 
-butter.api.start {port: 3001, host: "0.0.0.0"}, (err, server) =>
\ No newline at end of file
+butter.api.start {port: 3001, host: "0.0.0.0"}, (err, server) =>
```
Comment (PartTimeLegend) [inline bin/api:9]: I see no change on this line. What am I missing?

File: bin/engine [hunk 2]
```diff
@@ -3,4 +3,4 @@
 # Buttercoin engine binary script
 # This binary will start a single instance of a Buttercoin trading engine
 
-console.log "engine: Buttercoin trading engine started"
\ No newline at end of file
+console.log "engine: Buttercoin trading engine started"
```
Comment (PartTimeLegend) [inline bin/engine:6]: I see no change on this line. What am I missing?

File: bin/front [hunk 3]
```diff
@@ -5,4 +5,4 @@
 
 Buttercoin = require('../lib/buttercoin')
 butter = new Buttercoin();
-butter.front.start {port: 3000, host: "0.0.0.0", apiEndpoint: "ws://localhost:3001"}, (err, server) =>
\ No newline at end of file
+butter.front.start {port: 3000, host: "0.0.0.0", apiEndpoint: "ws://localhost:3001"}, (err, server) =>
```
Comment (PartTimeLegend) [inline bin/front:8]: I see no change on this line. What am I missing?

File: doc/CODING_GUIDELINES.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+= Coding Styleguide =
+
+Goal of this document is to end up with both readable and editable code.
+
+== Generic ==
+
+1. No tabs.
```
Comment (PartTimeLegend) [inline doc/CODING_GUIDELINES.md:7]: I actually prefer tabs to spaces. They are in my opinion faster to enter and allow for uniform alignment of code.

File: experimental/skiplist.js [hunk 5]
```diff
@@ -233,17 +234,14 @@ SkipList.prototype =  {
...
+                else {
                     n = n.d;
                 }
             }
 
         }
-    },
-
-    
+    }
```
Comment (PartTimeLegend) [inline experimental/skiplist.js:243]: Comma removed?

File: lib/api.coffee [hunk 6]
```diff
@@ -40,4 +40,4 @@ module.exports = class API
 
       ws.send 'i am api server ' + process.pid
 
-    callback null, wss
\ No newline at end of file
+    callback null, wss
```
Comment (PartTimeLegend) [inline lib/api.coffee:43]: What changed?

File: lib/engine.coffee [hunk 7]
```diff
@@ -39,4 +39,4 @@ module.exports = class Engine
       throw Error("Unknown Operation Type")
 
   flush: =>
-    @transaction_log.flush()
\ No newline at end of file
+    @transaction_log.flush()
```
Comment (PartTimeLegend) [inline lib/engine.coffee:42]: Another what changed?

File: lib/front.coffee [hunk 8]
```diff
@@ -111,4 +111,4 @@ module.exports = class Front
         wsClient.on 'message', (message) ->
           logger.info('front ' + process.pid +  ' received message: ' + message);
 
-          callback null, server
\ No newline at end of file
+          callback null, server
```
Comment (PartTimeLegend) [inline lib/front.coffee:114]: Again what changed?

File: bin/api (hunk 1 above)
Comment (RevCBH) [inline bin/api:9]: Ending the last line in a file with an newline character is generally regarded as best practice.

File: doc/CODING_GUIDELINES.md (hunk 4 above)
Comment (RevCBH) [inline doc/CODING_GUIDELINES.md:7]: A good editor should provide support for softtabs (i.e. emulating tabs with spaces). The problem with actual tab characters is that different editors will display tabs at different widths, which can break intended layout. For example, *nix defaults to 8 spaces per tab whereas Windows defaults to 4.

Comment (enki) [inline doc/CODING_GUIDELINES.md:7]: agreed with revcbh

File: experimental/skiplist.js (hunk 5 above)
Comment (RevCBH) [inline experimental/skiplist.js:243]: Looks like a trailing comma. This is valid JS but generally considered poor form: 
`{foo: 1, bar: 2, }`
The prefered form is:
`{foo: 1, bar: 2}`

File: lib/api.coffee (hunk 6 above)
Comment (RevCBH) [inline lib/api.coffee:43]: Newline character added at end of file

**Comment by RevCBH** [comment]
I think these changes are good. I know the tab thing requires more discussion, but it's easy to switch back.

**Pull request merged**